### ISSUE TITLE
BadParcelableException when unmarshalling Parcelable extending class

### ISSUE THREAD

**Issue by menny** [robolectric/robolectric#2398]
In the case where `class X extends B` and `abstract class B implements Parcelable`, Robolectric will fail to unmarshal class X.

I have an example project that shows this:
https://github.com/menny/auto-value-parcel-robolectric/tree/master

`MainActivity` has three static method:
1. `testParcel` - which shows [AutoValue-Parcel](https://github.com/rharter/auto-value-parcel) usage.
2. `testParcel2` - which creates the same class hierarchy manually.
3. `testParcelC` - which has a single class that declare the `Parcelable` interface and implements it.

When running on device, all three methods run without any exception, or errors.
When running a robolectric unit-test, `testParcel` and `testParcel2` will fail with `BadParcelableException`, which `testParcelC` passes.

The exception stack-trace is:

```
android.os.BadParcelableException: IllegalAccessException when unmarshalling: com.example.menny.roboautovalue.AutoValue_TestModel
    at android.os.Parcel.readParcelableCreator(Parcel.java:2099)
    at android.os.Parcel.readParcelable(Parcel.java:2055)
    at com.example.menny.roboautovalue.TestModelTest.testParcel(TestModelTest.java:22)
```

**Comment by menny** [comment]
@jongerrish @erd any ideas?

**Comment by menny** [comment]
@erd @jongerrish I have a fix PR: https://github.com/robolectric/robolectric/pull/2400
But I think it is not an elegant solution, a better solution would be to somehow make JDK8 `Field#get` _not_ do that caller's class verification:
```
@CallerSensitive
    public Object get(Object obj)
        throws IllegalArgumentException, IllegalAccessException
    {
        if (!override) {
            if (!Reflection.quickCheckMemberAccess(clazz, modifiers)) {
                Class<?> caller = Reflection.getCallerClass();
                checkAccess(caller, clazz, obj, modifiers);
            }
        }
        return getFieldAccessor(obj).get(obj);
    }
```
The check for caller-class does not exist in Android's JVM.

If you have pointers to how to do it, please let me know and I will revise this PR.

**Comment by jongerrish** [comment]
I'm wondering if we should be loading the code from the android sdk, I think currently we don't. Check InstrumentationConfiguration.shouldAquire() it seems very inconsistent on what gets instrumented / loaded through that classloader. i think we should be loading all android code, but java sdk stuff comes from the main classloader I think. We should probably look at fixing that, it would fix this issue too... not sure on the classloader restrictions of loading from java.lang.* though.

**Comment by menny** [comment]
@jongerrish what do you suggest I try out?

**Comment by jongerrish** [comment]
Check out InstrumentationConfiguration I think is the class and the method
shouldAquire() this determines whether the class is loaded from the
InstrumentationClassloader or the system class loader. If you can just
perform a little test and make sure this case is true for the class in
question it means we'd be loading the version of that class that is in the
Android SDK jar (rather than from the jdk you are using on the desk top)

Long term I want to make it so Robolectric prefers all the classes in the
Android jar as it makes for a higher fidelity environment, right now it's
not possible as there is not a clean separation between the system
environment and the test runtime environment but making an exception for
this one class could work to solve your problem.

**Comment by menny** [comment]
@jongerrish I've ran the tests with acquiring `java.lng.reflect.Field`. This is the error I get:
```
Tests run: 7, Failures: 0, Errors: 7, Skipped: 0, Time elapsed: 0 sec <<< FAILURE!
testInheritance[16](org.robolectric.shadows.ShadowAbsSeekBarTest)  Time elapsed: 0 sec  <<< ERROR!
java.util.ServiceConfigurationError: org.robolectric.internal.ShadowProvider: Provider org.robolectric.Shadows not a subtype
	at java.util.ServiceLoader.fail(ServiceLoader.java:239)
	at java.util.ServiceLoader.access$300(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:376)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at org.robolectric.internal.bytecode.InstrumentationConfiguration$Builder.build(InstrumentationConfiguration.java:151)
	at org.robolectric.RobolectricTestRunner.createClassLoaderConfig(RobolectricTestRunner.java:137)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:191)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:56)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:193)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:52)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:191)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:42)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:184)
	at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:159)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:236)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:24)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:193)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:52)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:191)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:42)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:184)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:236)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray2(ReflectionUtils.java:208)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:158)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:86)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:95)
```

I took this stack-trace from sunfire's output files.

**Comment by jongerrish** [comment]
It's still not the root cause exception

**Comment by menny** [comment]
Where can see the root cause?

**Comment by jongerrish** [comment]
Just patched this in, the root cause when I add special consideration for Field:-

  public boolean shouldAcquire(String name) {
    if (name.equals("java.lang.reflect.Field")) {
      return true;
    }

is:-
Caused by: java.lang.SecurityException: Prohibited package name: java.lang.reflect
	at java.lang.ClassLoader.preDefineClass(ClassLoader.java:661)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:797)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:643)
	at org.robolectric.internal.bytecode.InstrumentingClassLoader.findClass(InstrumentingClassLoader.java:177)

Wonder if there is any way around that?

**Comment by menny** [comment]
@jongerrish I found a very unwanted side-affect of my PR: since the bytecode changed, JaCoCo complains that `Execution data for class X does not match.`:
```
[ant:jacocoReport] Classes in bundle 'app' do no match with execution data. For report generation the same class files must be used as at runtime.
```
and as a result the code coverage drops! I guess that makes my PR unattractive. 

Do you have any suggestions for other ways to fix this issue?

**Comment by GrahamRogers** [comment]
Is there an update on this? It's blocking us from using auto-value-parcel in our project.

**Comment by menny** [comment]
I was able to overcome this issue with a custom shadow:
```
@Implements(Parcel.class)
public class ShadowParcelWithCreatorAccess extends ShadowParcel {
    @RealObject
    private Parcel realObject;

    @Implementation
    @HiddenApi
    public final Parcelable.Creator<?> readParcelableCreator(ClassLoader loader) throws ClassNotFoundException, NoSuchFieldException, IllegalAccessException {
        //note: calling `readString` will also consume the string, and increment the data-pointer.
        //which is exactly what we need, since we do not call the real `readParcelableCreator`.
        final String name = realObject.readString();

        ClassLoader parcelableClassLoader = (loader == null ? getClass().getClassLoader() : loader);
        Class<?> parcelableClass = Class.forName(name, false /* initialize */, parcelableClassLoader);
        Field field = parcelableClass.getField("CREATOR");
        //this is a fix for JDK8<->Android VM incompatibility:
        //Apparently, JDK will not allow access to a public field if its
        //class is not visible (private or package-private) from the call-site.
        field.setAccessible(true);

        return (Parcelable.Creator<?>) field.get(null/*static method does not have an instance*/);
    }
}
```
there is an obvious con to this approach: it does not run the real Android's `readParcelableCreator` implementation. But I don't think it's a huge issue.

I'll provide a PR shortly.

**Comment by xian** [comment]
Fixed by #2400.
